tools/generate_config.py

```diff
@@ -0,0 +1,326 @@
+
+    header.write('#include <inttypes.h>\n');
+    header.write('#include <stdbool.h>\n');
+    header.write('#include <stddef.h>\n');
+    header.write('#include <elf.h>\n\n');
+
+    header.write('#define SGXLKL_ENCLAVE_CONFIG_VERSION 1UL\n\n');
+
+    num_settings = 0
+    for typename, typedef in root['definitions'].items():
+      if (typename.startswith('sgxlkl_')):
+        if 'enum' in typedef:
+          names = typedef['c_enum'] if 'c_enum' in typedef else typedef['enum']
+          header.write('typedef enum\n{\n')
+          i = 0
+          num_vals = len(names)
+          for value in names:
+            header.write('    %s = %d' % (value.upper(), i))
+            if i < num_vals - 1: header.write(',')
+            if 'description' in typedef:
+              header.write(' /* %s */' % typedef['description'])
+            header.write('\n')
+            i += 1
+          header.write('} %s;\n\n' % typename)
+
+          if 'c_enum' in typedef:
+            header.write('const char* %s_to_string(%s e);\n' % (typename, typename))
+            header.write('%s string_to_%s(const char *e);\n\n' % (typename, typename))
+        else:
+          header.write('typedef struct %s\n{\n' % (typename[:-2]))
+          for name, jtype in typedef['properties'].items():
+            if name == 'format_version':
+              continue
+            elif (need_size_var(jtype)):
+              var_name = 'num_' + name;
+              if name == 'key':
+                var_name = 'key_len'
+              header.write('    size_t %s;\n' % var_name)
+            header.write('    %s %s%s;\n' % (pre_type(jtype), name, post_type(jtype)))
+            num_settings += 1
+          num_settings -= 1
+          header.write('} %s;\n\n' % typename)
+
+    header.write('extern const sgxlkl_enclave_config_t sgxlkl_default_enclave_config;\n\n');
+
+    header.write(
+      "typedef struct {\n"
+      "    char* scope;\n"
+      "    char* type;\n"
+      "    char* description;\n"
+      "    char* default_value;\n"
+      "    char* override_var;\n"
+      "} sgxlkl_enclave_setting_t;\n\n");
+
+    header.write("extern const sgxlkl_enclave_setting_t sgxlkl_enclave_settings[%d];\n\n" % num_settings);
+
+    header.write('#endif /* SGXLKL_ENCLAVE_CONFIG_H */');
+
+source_includes = """
+#ifdef SGXLKL_ENCLAVE
+#include <enclave/enclave_util.h>
+#include <enclave/oe_compat.h>
+#define FAIL sgxlkl_fail
+#else
+#include <host/sgxlkl_util.h>
+#include <string.h>
+#define FAIL sgxlkl_host_fail
+#endif
+"""
+
+def generate_source(schema_file_name, root, args):
+  with open(str(args.source), "w") as source:
+    source.write('/* Automatically generated from %s; do not modify. */\n' % schema_file_name);
+
+    source.write(source_includes)
+    source.write('\n')
+
+    source.write('#include "%s"\n\n' % args.header);
+
+    # enum conversions
+    for typename, typedef in root['definitions'].items():
+      if 'enum' in typedef:
```

Comment:
Why not avoid indentation by doing `if 'enum' not in typedef and 'c_enum' not in typedef: continue`?

Reply:
Because it doesn't make a difference + it's the same branch structure as in the other functions.

Reply:
Of course it makes a difference, less indentation == better for small screens and during split-editor reviewing.

Reply:
Except it looks like it has different semantics than the other functions, so you'll waste time looking for bugs where there aren't any. I'm more than happy to have you take a pass over it to make it more presentable to the Python community :-)